File: npmjs/pray-cli/src/manifest/format-manifest.ts
import { PrayError } from "../errors.js";
import type { ResolvedProject } from "../resolve/types.js";
import {
  bindLocalEntry,
  bindPackageEntry,
  exportKindMatchesRole,
  newDestinationTarget,
  packageBound,
  packageRoles,
} from "./destination.js";
import { serializeRecommended, targetHasExtras } from "./format-serialize.js";
import { parseManifestText } from "./parser.js";
import type {
  ExportRole,
  Manifest,
  ManifestLocal,
  ManifestPackage,
  ManifestTarget,
} from "./types.js";
import { canonicalManifest, manifestToJson } from "./types.js";

export interface PackageFormatHint {
  roles: ExportRole[];
  filePath?: string;
  /** Export names that must stay explicit after migration when a role is ambiguous. */
  exports?: string[];
}

export function usesDestinationDsl(manifest: Manifest): boolean {
  return (
    manifest.targets.some(
      (target) =>
        (target.scoped ?? false) || (target.mode ?? "legacy") !== "legacy",
    ) ||
    manifest.packages.some((entry) => packageBound(entry) || entry.file) ||
    manifest.local.some((entry) => entry.bound ?? false)
  );
}

function hasMigratableLegacyTargets(manifest: Manifest): boolean {
  return manifest.targets.some(
    (target) =>
      !(target.scoped ?? false) &&
      (target.mode ?? "legacy") === "legacy" &&
      (target.outputs.length > 0 || target.skills.length > 0),
  );
}

export function classifyFormatHints(
  project: ResolvedProject,
): Map<string, PackageFormatHint> {
  const hints = new Map<string, PackageFormatHint>();
  for (const packageEntry of project.packages) {
    const roles: ExportRole[] = [];
    let filePath = packageEntry.declaration.file;
    for (const exportName of packageEntry.selectedExports) {
      const exportEntry = packageEntry.spec.exports.get(exportName);
      if (!exportEntry) {
        continue;
      }
      for (const role of ["fragment", "folder", "file"] as ExportRole[]) {
        if (
          exportKindMatchesRole(exportEntry.kind, role) &&
          !roles.includes(role)
        ) {
          roles.push(role);
        }
      }
      if (!filePath && exportEntry.kind === "file") {
        filePath = exportEntry.defaultPath ?? exportName;
      }
    }
    const exports = ambiguousExportsForRoles(
      packageEntry.selectedExports,
      packageEntry.spec.exports,
      roles,
    );
    hints.set(packageEntry.declaration.name, { roles, filePath, exports });
  }
  return hints;
}

function ambiguousExportsForRoles(
  selectedExports: string[],
  exports: Map<string, { kind: string }>,
  roles: ExportRole[],
): string[] {
  const ambiguous: string[] = [];
  for (const role of roles) {
    const matching = selectedExports.filter((exportName) => {
      const exportEntry = exports.get(exportName);
      return (
        exportEntry !== undefined &&
        exportKindMatchesRole(exportEntry.kind, role)
      );
    });
    if (matching.length > 1) {
      for (const exportName of matching) {
        if (!ambiguous.includes(exportName)) {
          ambiguous.push(exportName);
        }
      }
    }
  }
  return ambiguous;
}

export function recommendManifest(
  manifest: Manifest,
  hints: Map<string, PackageFormatHint>,
): Manifest {
  // Top-level file: bindings alone must not skip migration of legacy target blocks.
  const recommended = hasMigratableLegacyTargets(manifest)
    ? migrateLegacyManifest(manifest, hints)
    : {
        ...manifest,
        packages: manifest.packages.map((entry) => ({
          ...entry,
          exports: [...entry.exports],
          targets: [...entry.targets],
          features: [...entry.features],
          groups: [...entry.groups],
          roles: [...(entry.roles ?? [])],
        })),
        local: manifest.local.map((entry) => ({ ...entry })),
        targets: manifest.targets.map((target) => ({
          ...target,
          outputs: [...target.outputs],
          skills: [...target.skills],
          commands: [...target.commands],
          rules: [...target.rules],
          entries: [...(target.entries ?? [])],
        })),
      };
  omitContextResolvedExports(recommended);
  omitDefaultSources(recommended);
  return recommended;
}

function omitContextResolvedExports(manifest: Manifest): void {
  for (const packageEntry of manifest.packages) {
    if (packageBound(packageEntry) && packageEntry.exports.length <= 1) {
      packageEntry.exports = [];
    }
  }
}

function packageNamespace(name: string): string | undefined {
  const separator = name.indexOf("/");
  return separator === -1 ? undefined : name.slice(0, separator);
}

function omitDefaultSources(manifest: Manifest): void {
  const soleSource =
    manifest.sources.length === 1 ? manifest.sources[0]?.name : undefined;
  const sourceNames = new Set(manifest.sources.map((source) => source.name));
  for (const packageEntry of manifest.packages) {
    const source = packageEntry.source;
    if (!source) {
      continue;
    }
    const matchesSole = soleSource === source;
    const matchesNamespace =
      packageNamespace(packageEntry.name) === source && sourceNames.has(source);
    if (matchesSole || matchesNamespace) {
      packageEntry.source = undefined;
    }
  }
}

export function formatRecommended(
  manifest: Manifest,
  hints: Map<string, PackageFormatHint>,
): string {
  const recommended = recommendManifest(manifest, hints);
  const text = serializeRecommended(recommended);
  const reparsed = parseManifestText(text);
  if (
    JSON.stringify(manifestToJson(reparsed)) !==
    JSON.stringify(manifestToJson(canonicalManifest(recommended)))
  ) {
    throw PrayError.manifest(
      "formatted Prayfile did not round-trip to an equivalent manifest",
    );
  }
  return text;
}

function migrateLegacyManifest(
  manifest: Manifest,
  hints: Map<string, PackageFormatHint>,
): Manifest {
  const next: Manifest = {
    prayfileVersion: manifest.prayfileVersion,
    sources: manifest.sources,
    targets: [],
    packages: manifest.packages.map((entry) => ({
      ...entry,
      exports: [...entry.exports],
      targets: [...entry.targets],
      features: [...entry.features],
      groups: [...entry.groups],
      roles: [...(entry.roles ?? [])],
    })),
    local: manifest.local.map((entry) => ({ ...entry })),
    render: manifest.render,
  };

  applyFormatHints(next.packages, hints);

  const composePaths = uniquePaths(
    manifest.targets.flatMap((target) =>
      target.outputs.map((path) => [path, target.name] as [string, string]),
    ),
  );
  const treePaths = uniquePaths(
    manifest.targets.flatMap((target) =>
      target.skills.map((path) => [path, target.name] as [string, string]),
    ),
  );

  for (const [path, targetNames] of composePaths) {
    const target = newDestinationTarget("compose", path);
    for (const local of localsForCompose(next.local)) {
      bindLocalEntry(target, local.path);
      const entry = next.local.find(
        (candidate) => candidate.path === local.path,
      );
      if (entry) {
        entry.bound = true;
      }
    }
    for (const packageEntry of packagesForRole(
      next.packages,
      "fragment",
      targetNames,
    )) {
      bindPackageEntry(target, packageEntry.name);
      markPackageBound(next.packages, packageEntry.name, "fragment");
    }
    next.targets.push(target);
  }

  for (const [path, targetNames] of treePaths) {
    const target = newDestinationTarget("tree", path);
    for (const packageEntry of packagesForRole(
      next.packages,
      "folder",
      targetNames,
    )) {
      bindPackageEntry(target, packageEntry.name);
      markPackageBound(next.packages, packageEntry.name, "folder");
    }
    next.targets.push(target);
  }

  for (const packageEntry of next.packages) {
    if (packageEntry.file) {
      packageEntry.bound = true;
      const roles = packageRoles(packageEntry);
      if (!roles.includes("file")) {
        roles.push("file");
      }
      packageEntry.roles = roles;
    }
  }
  for (const local of next.local) {
    if (local.bound) {
      local.position = "after";
    }
  }

  for (const target of manifest.targets) {
    if (targetHasExtras(target)) {
      next.targets.push({
        name: target.name,
        outputs: [],
        skills: [],
        commands: target.commands,
        rules: target.rules,
        maxBytes: target.maxBytes,
        mode: "legacy",
        scoped: false,
        entries: [],
      });
    }
  }

  return next;
}

function applyFormatHints(
  packages: ManifestPackage[],
  hints: Map<string, PackageFormatHint>,
): void {
  for (const packageEntry of packages) {
    const hint = hints.get(packageEntry.name);
    if (hint) {
      const roles = packageRoles(packageEntry);
      for (const role of hint.roles) {
        if (!roles.includes(role)) {
          roles.push(role);
        }
      }
      packageEntry.roles = roles;
      if (!packageEntry.file) {
        packageEntry.file = hint.filePath;
      }
      if (
        packageEntry.exports.length === 0 &&
        hint.exports &&
        hint.exports.length > 0
      ) {
        packageEntry.exports = [...hint.exports];
      }
    }
    if (packageEntry.file && !packageRoles(packageEntry).includes("file")) {
      const roles = packageRoles(packageEntry);
      roles.push("file");
      packageEntry.roles = roles;
    }
  }
}

function uniquePaths(
  items: Array<[string, string]>,
): Array<[string, Set<string>]> {
  const map = new Map<string, Set<string>>();
  for (const [path, targetName] of items) {
    const set = map.get(path) ?? new Set<string>();
    set.add(targetName);
    map.set(path, set);
  }
  return [...map.entries()].sort(([left], [right]) =>
    left.localeCompare(right),
  );
}

function localsForCompose(locals: ManifestLocal[]): ManifestLocal[] {
  const before: ManifestLocal[] = [];
  const after: ManifestLocal[] = [];
  for (const local of locals) {
    if (local.bound) {
      continue;
    }
    if (local.position === "before") {
      before.push(local);
    } else {
      after.push(local);
    }
  }
  return [...before, ...after];
}

function packagesForRole(
  packages: ManifestPackage[],
  role: ExportRole,
  targetNames: Set<string>,
): ManifestPackage[] {
  return packages.filter((packageEntry) => {
    if (packageEntry.file) {
      return false;
    }
    if (
      packageEntry.targets.length > 0 &&
      !packageEntry.targets.some((name) => targetNames.has(name))
    ) {
      return false;
    }
    return packageRoles(packageEntry).includes(role);
  });
}

function markPackageBound(
  packages: ManifestPackage[],
  name: string,
  role: ExportRole,
): void {
  const packageEntry = packages.find((entry) => entry.name === name);
  if (!packageEntry) {
    return;
  }
  packageEntry.bound = true;
  const roles = packageRoles(packageEntry);
  if (!roles.includes(role)) {
    roles.push(role);
  }
  packageEntry.roles = roles;
}

export type { ManifestTarget };

File: crates/pray-core/src/format_manifest.rs
use crate::destination::{
    bind_local_entry, bind_package_entry, export_kind_matches_role, new_destination_target,
};
use crate::format_serialize::{serialize_recommended, target_has_extras};
use crate::manifest::{
    DestinationMode, ExportRole, Manifest, ManifestLocal, ManifestPackage, ManifestTarget,
};
use crate::resolve::ResolvedProject;
use crate::{PrayError, PrayResult};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Debug, Clone, Default)]
pub struct PackageFormatHint {
    pub roles: Vec<ExportRole>,
    pub file_path: Option<String>,
    /// Export names that must stay explicit after migration when a role is ambiguous.
    pub exports: Vec<String>,
}

pub fn uses_destination_dsl(manifest: &Manifest) -> bool {
    manifest
        .targets
        .iter()
        .any(|target| target.scoped || target.mode != DestinationMode::Legacy)
        || manifest
            .packages
            .iter()
            .any(|package| package.bound || package.file.is_some())
        || manifest.local.iter().any(|local| local.bound)
}

fn has_migratable_legacy_targets(manifest: &Manifest) -> bool {
    manifest.targets.iter().any(|target| {
        !target.scoped
            && target.mode == DestinationMode::Legacy
            && (!target.outputs.is_empty() || !target.skills.is_empty())
    })
}

pub fn classify_format_hints(project: &ResolvedProject) -> BTreeMap<String, PackageFormatHint> {
    let mut hints = BTreeMap::new();
    for package in &project.packages {
        let mut roles = Vec::new();
        let mut file_path = package.declaration.file.clone();
        for export_name in &package.selected_exports {
            let Some(export) = package.spec.exports.get(export_name) else {
                continue;
            };
            for role in [ExportRole::Fragment, ExportRole::Folder, ExportRole::File] {
                if export_kind_matches_role(&export.kind, role) && !roles.contains(&role) {
                    roles.push(role);
                }
            }
            if file_path.is_none() && export.kind == "file" {
                file_path = export
                    .default_path
                    .clone()
                    .or_else(|| Some(export_name.clone()));
            }
        }
        let exports =
            ambiguous_exports_for_roles(&package.selected_exports, &package.spec.exports, &roles);
        hints.insert(
            package.declaration.name.clone(),
            PackageFormatHint {
                roles,
                file_path,
                exports,
            },
        );
    }
    hints
}

fn ambiguous_exports_for_roles(
    selected_exports: &[String],
    exports: &BTreeMap<String, crate::package_spec::PackageExport>,
    roles: &[ExportRole],
) -> Vec<String> {
    let mut ambiguous = Vec::new();
    for role in roles {
        let matching: Vec<String> = selected_exports
            .iter()
            .filter(|export_name| {
                exports
                    .get(*export_name)
                    .is_some_and(|export| export_kind_matches_role(&export.kind, *role))
            })
            .cloned()
            .collect();
        if matching.len() > 1 {
            for export_name in matching {
                if !ambiguous.contains(&export_name) {
                    ambiguous.push(export_name);
                }
            }
        }
    }
    ambiguous
}

pub fn recommend_manifest(
    manifest: &Manifest,
    hints: &BTreeMap<String, PackageFormatHint>,
) -> Manifest {
    // Top-level file: bindings alone must not skip migration of legacy target blocks.
    let mut recommended = if has_migratable_legacy_targets(manifest) {
        migrate_legacy_manifest(manifest, hints)
    } else {
        manifest.clone()
    };
    omit_context_resolved_exports(&mut recommended);
    omit_default_sources(&mut recommended);
    recommended
}

fn omit_context_resolved_exports(manifest: &mut Manifest) {
    for package in &mut manifest.packages {
        if package.bound && package.exports.len() <= 1 {
            package.exports.clear();
        }
    }
}

fn package_namespace(name: &str) -> Option<&str> {
    name.split_once('/').map(|(namespace, _)| namespace)
}

fn omit_default_sources(manifest: &mut Manifest) {
    let sole_source = (manifest.sources.len() == 1).then(|| manifest.sources[0].name.clone());
    let source_names: BTreeSet<&str> = manifest
        .sources
        .iter()
        .map(|source| source.name.as_str())
        .collect();
    for package in &mut manifest.packages {
        let Some(source) = package.source.as_deref() else {
            continue;
        };
        let matches_sole = sole_source.as_deref() == Some(source);
        let matches_namespace =
            package_namespace(&package.name) == Some(source) && source_names.contains(source);
        if matches_sole || matches_namespace {
            package.source = None;
        }
    }
}

pub fn format_recommended(
    manifest: &Manifest,
    hints: &BTreeMap<String, PackageFormatHint>,
) -> PrayResult<String> {
    let recommended = recommend_manifest(manifest, hints);
    let text = serialize_recommended(&recommended);
    let reparsed = crate::manifest::parse_manifest(&text)?;
    if reparsed.canonicalized() != recommended.canonicalized() {
        return Err(PrayError::Manifest(
            "formatted Prayfile did not round-trip to an equivalent manifest".to_string(),
        ));
    }
    Ok(text)
}

fn migrate_legacy_manifest(
    manifest: &Manifest,
    hints: &BTreeMap<String, PackageFormatHint>,
) -> Manifest {
    let mut next = Manifest {
        prayfile_version: manifest.prayfile_version.clone(),
        sources: manifest.sources.clone(),
        targets: Vec::new(),
        packages: manifest.packages.clone(),
        local: manifest.local.clone(),
        symbols: manifest.symbols.clone(),
        render: manifest.render.clone(),
    };

    apply_format_hints(&mut next.packages, hints);

    let compose_paths = unique_paths(manifest.targets.iter().flat_map(|target| {
        target
            .outputs
            .iter()
            .map(move |path| (path.clone(), target.name.clone()))
    }));
    let tree_paths = unique_paths(manifest.targets.iter().flat_map(|target| {
        target
            .skills
            .iter()
            .map(move |path| (path.clone(), target.name.clone()))
    }));

    for (path, target_names) in &compose_paths {
        let mut target = new_destination_target(DestinationMode::Compose, path);
        for local in locals_for_compose(&next.local) {
            bind_local_entry(&mut target, &local.path);
            if let Some(entry) = next
                .local
                .iter_mut()
                .find(|candidate| candidate.path == local.path)
            {
                entry.bound = true;
            }
        }
        for package in packages_for_role(&next.packages, ExportRole::Fragment, target_names) {
            bind_package_entry(&mut target, &package.name);
            mark_package_bound(&mut next.packages, &package.name, ExportRole::Fragment);
        }
        next.targets.push(target);
    }

    for (path, target_names) in &tree_paths {
        let mut target = new_destination_target(DestinationMode::Tree, path);
        for package in packages_for_role(&next.packages, ExportRole::Folder, target_names) {
            bind_package_entry(&mut target, &package.name);
            mark_package_bound(&mut next.packages, &package.name, ExportRole::Folder);
        }
        next.targets.push(target);
    }

    for package in &mut next.packages {
        if package.file.is_some() {
            package.bound = true;
            if !package.roles.contains(&ExportRole::File) {
                package.roles.push(ExportRole::File);
            }
        }
    }
    for local in &mut next.local {
        if local.bound {
            // Compose entry order replaces legacy position keywords.
            local.position = "after".to_string();
        }
    }

    for target in &manifest.targets {
        if target_has_extras(target) {
            next.targets.push(ManifestTarget {
                name: target.name.clone(),
                outputs: Vec::new(),
                skills: Vec::new(),
                commands: target.commands.clone(),
                rules: target.rules.clone(),
                max_bytes: target.max_bytes,
                mode: DestinationMode::Legacy,
                scoped: false,
                entries: Vec::new(),
            });
        }
    }

    next
}

fn apply_format_hints(
    packages: &mut [ManifestPackage],
    hints: &BTreeMap<String, PackageFormatHint>,
) {
    for package in packages {
        if let Some(hint) = hints.get(&package.name) {
            for role in &hint.roles {
                if !package.roles.contains(role) {
                    package.roles.push(*role);
                }
            }
            if package.file.is_none() {
                package.file = hint.file_path.clone();
            }
            if package.exports.is_empty() && !hint.exports.is_empty() {
                package.exports = hint.exports.clone();
            }
        }
        if package.file.is_some() && !package.roles.contains(&ExportRole::File) {
            package.roles.push(ExportRole::File);
        }
    }
}

fn unique_paths(items: impl Iterator<Item = (String, String)>) -> Vec<(String, BTreeSet<String>)> {
    let mut map: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
    for (path, target_name) in items {
        map.entry(path).or_default().insert(target_name);
    }
    map.into_iter().collect()
}

fn locals_for_compose(locals: &[ManifestLocal]) -> Vec<ManifestLocal> {
    let mut before = Vec::new();
    let mut after = Vec::new();
    for local in locals {
        if local.bound {
            continue;
        }
        match local.position.as_str() {
            "start" | "before" => before.push(local.clone()),
            _ => after.push(local.clone()),
        }
    }
    before.extend(after);
    before
}

fn packages_for_role(
    packages: &[ManifestPackage],
    role: ExportRole,
    target_names: &BTreeSet<String>,
) -> Vec<ManifestPackage> {
    packages
        .iter()
        .filter(|package| {
            if package.file.is_some() {
                return false;
            }
            if !package.targets.is_empty()
                && !package
                    .targets
                    .iter()
                    .any(|name| target_names.contains(name))
            {
                return false;
            }
            package.roles.contains(&role)
        })
        .cloned()
        .collect()
}

fn mark_package_bound(packages: &mut [ManifestPackage], name: &str, role: ExportRole) {
    if let Some(package) = packages.iter_mut().find(|package| package.name == name) {
        package.bound = true;
        if !package.roles.contains(&role) {
            package.roles.push(role);
        }
    }
}
